src/Microsoft.CodeQuality.Analyzers.Exp/Core/Reliability/DisposeObjectsBeforeLosingScope.cs

```diff
@@ -0,0 +1,141 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Operations.ControlFlow;
+using Microsoft.CodeAnalysis.Operations.DataFlow;
+using Microsoft.CodeAnalysis.Operations.DataFlow.DisposeAnalysis;
+using Microsoft.CodeAnalysis.Operations.DataFlow.NullAnalysis;
+using Microsoft.CodeAnalysis.Operations.DataFlow.PointsToAnalysis;
+
+namespace Microsoft.CodeQuality.Analyzers.Exp.Reliability
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class DisposeObjectsBeforeLosingScope : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA2000";
+
+        private static readonly string[] s_disposeOwnershipTransferLikelyTypes = new string[]
```

Comment:
Can you file a follow up issue here for creating a way for a user to mark that their method/constructor takes ownership of the disposable?

Reply:
Good idea. Filed https://github.com/dotnet/roslyn-analyzers/issues/1587